### PR TITLE
Correct back propagation

### PR DESCRIPTION
**Pull request by willmarks**
Adds a checkpoint node that is used to terminate back propagation so that each layer can be fully computed.

Its also stupid fast now... hopefully not because of a bug

### CODE REVIEW

File: src/engine/node.rs
Comment (willmarks) [inline src/engine/node.rs:1]: Renamed Value to Node.

File: src/engine/node.rs [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+pub fn checkpoint(node: usize, name: &str, state: &mut State) -> usize {
+    let out = Node {
+        data: state[node].data,
+        name: format!("Checkpoint({})", name),
+        _prev: (Some(node), None),
+        _op: Op::Pass,
+        ..Default::default()
+    };
+    out.push_on(state)
+}
```
Comment (willmarks) [inline src/engine/node.rs:50]: Checkpoint is a simple pass-through node that allows us to break back propagation at critical places.

File: src/engine/node.rs [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+
+/**
+ * Computes the backward pass until it reaches a checkpoint
+ */
+fn _back(node: usize, state: &mut State) {
+    match state[node]._op {
+        Op::Pass => (), // println!("- Stopping: {}", state[node].name),
+        _ => back(node, state)
+    }
+}
```
Comment (willmarks) [inline src/engine/node.rs]: recursive back propagation stops at checkpoints.

File: src/mind/neuron.rs [hunk 3]
```diff
@@ -2,53 +2,65 @@ use std::fmt;
...
+                }
+                None => {
+                    act = Some(mul_res);
+                }
             }
         }
-    }
 
-    tanh(add(act.unwrap(), neuron.bias, nodes), nodes)
+        checkpoint(
```
Comment (willmarks) [inline src/mind/neuron.rs:52]: wrapping a neuron's apply in a checkpoint prevents the neuron from being affected before all upstream back propagation is finished and back is called directly on this neurons checkpoint.

**Comment by AnthonyDario** [comment]
Merged this without reviewers approval!

**Pull request merged**